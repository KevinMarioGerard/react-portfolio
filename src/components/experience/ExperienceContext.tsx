import * as React from 'react';

export type SelectedState = string | null;

type ExperienceContextValue = {
    /**
     * Selected experience tab state
     */
    selected: SelectedState;

    /**
     * Set selected experience tab state
     */
    setSelected: (state: string) => void;
};

const ExperienceContext = React.createContext<ExperienceContextValue>({
    selected: null,
    setSelected: () => null,
});

export default ExperienceContext;
