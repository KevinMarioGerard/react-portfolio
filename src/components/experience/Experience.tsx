import * as React from 'react';
import ExperienceContext, { SelectedState } from './ExperienceContext';
import { ExperienceTabs, ExperienceTab } from './ExperienceTabs';
import ExperienceItem from './ExperienceItem';

type ExperienceContentProps = {
    /**
     * id mapped to experience tab.
     */
    id: string;
};

const ExperienceContent: React.FC<ExperienceContentProps> = ({ children, id }) => {
    const { selected } = React.useContext(ExperienceContext);

    return (
        <>
            {id === selected ? (
                <div className="experience__content__container" id={id}>
                    {children}
                </div>
            ) : null}
        </>
    );
};

ExperienceContent.displayName = 'Experience.Content';

type Statics = {
    Content: typeof ExperienceContent;

    Tabs: typeof ExperienceTabs;

    Tab: typeof ExperienceTab;

    Item: typeof ExperienceItem;
};

const Experience: React.FC & Statics = ({ children }) => {
    const [selected, setSelected] = React.useState<SelectedState>(null);

    return (
        <div className="experience__container">
            <ExperienceContext.Provider value={{ selected, setSelected }}>
                {children}
            </ExperienceContext.Provider>
        </div>
    );
};

Experience.displayName = 'Experience';

Experience.Content = ExperienceContent;
Experience.Item = ExperienceItem;
Experience.Tabs = ExperienceTabs;
Experience.Tab = ExperienceTab;

export default Experience;
