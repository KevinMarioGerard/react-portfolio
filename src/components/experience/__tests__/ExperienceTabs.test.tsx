import { render, fireEvent, getByText } from '@testing-library/react';
import Experience from '../Experience';
import * as React from 'react';
import ExperienceContext, { SelectedState } from '../ExperienceContext';

const TestExperienceTabs = () => {
    const [selected, setSelected] = React.useState<SelectedState>(null);

    return (
        <ExperienceContext.Provider value={{ selected, setSelected }}>
            <Experience.Tabs initial="test1">
                <Experience.Tab id="test1" label="Tab 1" />
                <Experience.Tab id="test2" label="Tab 2" />
                <Experience.Tab id="test3" label="Tab 3" />
            </Experience.Tabs>
        </ExperienceContext.Provider>
    );
};

describe('Experience Tabs Component tests', () => {
    it('should render without errors', () => {
        expect(() => {
            render(<TestExperienceTabs />);
        }).not.toThrow();
    });

    it('should match snapshot with initial selection', () => {
        const { container } = render(<TestExperienceTabs />);

        expect(container).toMatchSnapshot();
    });

    it('should update content on click', () => {
        const { container } = render(<TestExperienceTabs />);

        fireEvent.click(getByText(container, 'Tab 3'));

        expect(container).toMatchSnapshot();
    });
});

describe('Experience Tab Component tests', () => {
    it('should set selected state on click', () => {
        let selected = null;
        const setSelected = jest.fn(value => {
            selected = value;
        });

        const { container } = render(
            <ExperienceContext.Provider value={{ selected, setSelected }}>
                <Experience.Tab id="test2" label="Tab 2" />
            </ExperienceContext.Provider>
        );

        expect(selected).toBeNull();

        fireEvent.click(getByText(container, 'Tab 2'));

        expect(setSelected).toHaveBeenCalledTimes(1);
        expect(selected).toBe('test2');
    });
});
