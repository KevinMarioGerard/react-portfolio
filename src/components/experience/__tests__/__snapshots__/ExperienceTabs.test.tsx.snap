// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Experience Tabs Component tests should match snapshot with initial selection 1`] = `
<div>
  <div
    class="experience__tabs"
  >
    <div
      class="experience__tabs__container"
    >
      <div
        class="experience__tab experience__tab--selected"
        id="test1"
      >
        <span>
          Tab 1
        </span>
      </div>
      <div
        class="experience__tab "
        id="test2"
      >
        <span>
          Tab 2
        </span>
      </div>
      <div
        class="experience__tab "
        id="test3"
      >
        <span>
          Tab 3
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Experience Tabs Component tests should update content on click 1`] = `
<div>
  <div
    class="experience__tabs"
  >
    <div
      class="experience__tabs__container"
    >
      <div
        class="experience__tab "
        id="test1"
      >
        <span>
          Tab 1
        </span>
      </div>
      <div
        class="experience__tab "
        id="test2"
      >
        <span>
          Tab 2
        </span>
      </div>
      <div
        class="experience__tab experience__tab--selected"
        id="test3"
      >
        <span>
          Tab 3
        </span>
      </div>
    </div>
  </div>
</div>
`;
